proto/authn/v1/authn.proto

```diff
@@ -0,0 +1,71 @@
+// Authn protocol
+syntax = "proto3";
+package xmtp.authn.v1;
+
+import "google/api/annotations.proto";
+
+option go_package = "./authn/v1";
+
+// ----------------------------------------------------------------------------
+// These objects are copied from xmtpjs.
+// Compile failure occurs with GoGo protobuf as optional fields are not 
+// supported.
+// Migrating to go
+// ----------------------------------------------------------------------------
+
+// Signature represents a generalized public key signature,
+// defined as a union to support cryptographic algorithm agility.
+message Signature {
+    message ECDSACompact {
+        bytes bytes = 1; // compact representation [ R || S ], 64 bytes
+        uint32 recovery = 2; // recovery bit
+    }
+    oneof union {
+        ECDSACompact ecdsa_compact = 1;
+    }
+}
+
+message Secp256k1Uncompresed {
+    // uncompressed point with prefix (0x04) [ P || X || Y ], 65 bytes
+    bytes bytes = 1; 
+}
+// PublicKey represents a generalized public key,
+// defined as a union to support cryptographic algorithm agility.
+message PublicKey {
+    uint64 timestamp = 1;
+    Signature signature = 2;
+    oneof union {
+        Secp256k1Uncompresed secp256k1_uncompressed = 3;
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+message ClientAuthRequest {
+    PublicKey identity_key = 1;
+    bytes auth_data_bytes = 3;
+    Signature auth_data_signature = 4;
+}
+  
+message ClientAuthResponse {
+    oneof union {
+        bytes token = 1;  // if auth was successful
+        string error_str = 2; // if auth was not successful
```

Comment:
Why can't we just use GRPC's native status codes, with custom messages, to pass error responses?

Reply:
Maybe we can, I'm not quite sure how that works. I'm still just playing with this, trying to figure out what I want to do.

At this point I'm thinking the token could really be just a random number, that we associated with the wallet when auth succeeds (this is the replacement for peer Id in peer store). The client will have to pass the token along with every request to the API and until the token is purged from the store (purgeLoop) the requests will be allowed. I think this would be an equivalent of what we had.

But then it gets interesting what happens when we purge the token and the connection is still active, somehow the client would have to re-engage the auth flow to get a new token. Makes me wonder if we couldn't just use the AuthData bundled as a token with every request to begin with. Then we don't have to worry about the protocol interplay and requests will be automatically portable between nodes too. Thoughts?

Reply:
To clarify the token would be produced by the client and it would sign a timestamp and wallet address using the identity key signed by the wallet. The server would just check the token on every request. If the token is too old or signatures don't check out, the request is rejected (probably using GRPC interceptors to implement this). That's it.

Reply:
That doesn't sound crazy to me. Little bit of extra overhead verifying all the signatures on each publish, but there's no network calls or anything like that. I like that it's stateless and portable.

Apps would be able to generate tokens infinitely into the future, but that was possible with the previous flow as well (so long as they kept the same peerID). Probably not our biggest worry. 

Wdyt @jazzz?

Reply:
We can reject tokens with future timestamps.

Reply:
Also, in general signature verification should be cheaper than signing (unless that changed with the EC world). So the important thing performance wise would probably be caching the token on the client side while it's valid.

Reply:
It's a case of stockpiling. You could generate them to the year 3000 and then use them at the right moment.

Reply:
Will definitely cache in the client. We'll need to agree on some validity window (60 mins?) and regenerate if expired.

Reply:
Looks like I was wrong https://www.ietf.org/proceedings/92/slides/slides-92-lwig-3.pdf, claims that EC verification is slower than EC signing :til:. Even so, if it becomes an issue, nodes can cache tokens that they've already seen as well, if it comes to that.

Reply:
We would have to make sure this whole thing serializes into a small enough size to fit in the HTTP Headers. Usually that gets cut off at 4kb or 8kb

Reply:
> It's a case of stockpiling. You could generate them to the year 3000 and then use them at the right moment.

Interesting, I'm not sure what would motivate a client to do that. It just increases a chance of someone stealing the token and impersonate them. I think rejecting future tokens can help making far future tokens less valuable. They would still have an "expiration window" sized period where it would be valid.

Reply:
> I'm not sure what would motivate a client to do that

Ya, most of the cases I can think of would be in the event of a malicious client...in which case we have bigger problems to worry about, since they can generate new tokens whenever they want.

Reply:
JWT auth with GRPC interceptors https://dev.to/techschoolguru/use-grpc-interceptor-for-authorization-with-jwt-1c5h

Reply:
> We would have to make sure this whole thing serializes into a small enough size to fit in the HTTP Headers. Usually that gets cut off at 4kb or 8kb

Yup, shouldn't be too bad, the signatures should be on the order of the key sizes, so I'd estimate good bit less than 1kb total.

Reply:
Few Questions:

- In this scheme how is the node providing the sybil resistance portion of Authn? Is each node checking a cache of wallets it has verified and then polling as needed?

- The PeerId previously was providing replay and PITM resistance as the authentication was performed per channel which was tied to a client EC Pair. Do we have a mechanism to tie this token to a particular channel? or minimally a client application? 

- I'm assuming were using TLS for client->gateway security. Are we pinning certificates?

Reply:
> In this scheme how is the node providing the sybil resistance portion of Authn? Is each node checking a cache of wallets it has verified and then polling as needed?

Think that is the idea. I would imagine the node would look up the wallet transaction history from the Alchemy API on first use of a wallet address, and all rate-limiting would be based on the wallet address and not the token.

> The PeerId previously was providing replay and PITM resistance as the authentication was performed per channel which was tied to a client EC Pair. Do we have a mechanism to tie this token to a particular channel? or minimally a client application?

GRPC Gateway is just a regular HTTP API, we have no other form of client authentication built in. This would be a bearer token. The token would be valid for N minutes, after which it would be rejected. This does mean that a compromised key would be usable by a PITM for N minutes.

Reply:
> Few Questions:
> 
> * In this scheme how is the node providing the sybil resistance portion of Authn? Is each node checking a cache of wallets it has verified and then polling as needed?

Not quite sure what you mean by sybil resistence here. The limits are tied to a wallet, Authn tied a wallet to a peerId so a request authorization would go peerId -> wallet -> limit. This scheme just skips the peerId bit and ties each requests to a wallet directly. It does that doing the same thing as authn did to tie the peerId to wallet. The downside we do the wallet validation on every request.

> * The PeerId previously was providing replay and PITM resistance as the authentication was performed per channel which was tied to a client EC Pair. Do we have a mechanism to tie this token to a particular channel? or minimally a client application?

There doesn't seem to be a way to tie incoming requests to channels in GRPC world. I'm not sure how to replicate that aspect here. I'm not sure we need to worry about MITM, that should be handled by TLS between the client and the server. The only replay protection with the Token is its expiration period, so if it does get stolen somehow (by a node?) that is an issue for a time. This seems on par with what JWT based authorization does.

We could try to emulate the peerID effect by having the client generate an EC pair, include the public key in the token and sign each request using the private key. That would complicate the message protocol quite a bit. We'd have to repackage each request type into some kind of signed structure. Is that worth it? Do we need to do better than common web auth does?

Reply:
Given that all nodes on the network already have the capability to send unlimited messages, and that this token is only used for sybil resistance/rate limiting of user clients, I'm not concerned about nodes stealing tokens. The horse has already left the barn there.

Reply:
@neekolas all sounds reasonable. 
 
 > We could try to emulate the peerID effect by having the client generate an EC pair,
 @mkobetic agreed I don't think this makes sense. I'm thinking forward to the work on malicious client prevention, where we might want a tight coupling of token to application. Non of this should stall this work. 
 
 I think this plan makes sense for the near horizon. As mechanisms such as postage become reality we can migrate to a more privacy preserving method to keeping the network secure.

Reply:
> You could generate them to the year 3000 and then use them at the right moment.

If we could include a challenge response, that would be helpful for mitigating future malicious clients . Not sure how feasible that is.

The server provides an input which must be included in the token. In an optimal world it would be unique to a session but I get this this would impact the portability across nodes.

Even if the challenge was a slowly changing network variable (Think block height), old tokens could be invalidated. There's still a window of compromise, but would likely remove the stockpiling issue as it would be less appealing to be storing T*|C| entries

Reply:
effectively establishing an upper and lower bound on the validity of tokens. This would mean when the rest of the malicious client mitigation work is completed we won't need to worry about the risk of legacy tokens.

Reply:
Isn't the timestamp basically a "slowly changing network variable"? I don't see stockpiling as an issue. There's no point for a valid client to stockpile tokens. There's nothing to gain from it other then increasing a chance of someone stealing them.

Reply:
> effectively establishing an upper and lower bound on the validity of tokens. This would mean when the rest of the malicious client mitigation work is completed we won't need to worry about the risk of legacy tokens.

The bounds I want to set with this design are
1. node_time < token.timestamp
2. token.timestamp < node_time + expiration(1h?)

So a token is only valid between token.timestamp and token.timestamp+expiration interpreted in node time.

Reply:
> There's no point for a valid client to stockpile tokens.

Agreed but I'm thinking in the context of a malicious node.

> Isn't the timestamp basically a "slowly changing network variable"?

Time is deterministic, and since the timestamp field is client supplied and untrustworthy theres no guarantee of when that token was generated. By adding a random variable to be included, we build confidence that the token was generated after the challenge was made public.

Reply:
There are other more substantial issues that need to be addressed before this is the limiting factor, so I'm happy to revisit it later. However having a "proof of life" style mechanic that establishes a lowerbound on when the token was generated would reduce risk of token abuse

Reply:
> Agreed but I'm thinking in the context of a malicious client.

At least right now, the malicious client would also have the `IdentityKey`s needed to generate new tokens on demand.